miniapps/nurbs/nurbs_naca_cmesh.cpp

```diff
@@ -0,0 +1,798 @@
+   output << "1 1 5 4" << endl;   // Bottom
+   output << "1 1 6 5" << endl;   // Bottom
+   output << "2 1 7 6" << endl;   // Inflow
+   output << "3 1 8 7" << endl;   // Top
+   output << "3 1 9 8" << endl;   // Top
+   output << "4 1 4 0" << endl;   // Outflow
+   output << "4 1 0 9" << endl;   // Outflow
+   output << "5 1 1 2" << endl;   // Foil section
+   output << "5 1 2 3" << endl;   // Foil section
+   output << "5 1 3 1" << endl;   // Foil section
+   output << endl;
+
+   // Edges
+   output <<"edges"<<endl;
+   output <<"15"<<endl;
+   output << "0 0 1"<<endl;
+   output << "1 1 2"<<endl;
+   output << "2 2 3"<<endl;
+   output << "3 3 1"<<endl;
+
+   output << "0 4 5"<<endl;
+   output << "1 5 6"<<endl;
+   output << "2 6 7"<<endl;
+   output << "3 7 8"<<endl;
+   output << "0 9 8"<<endl;
+
+   output << "4 0 4"<<endl;
+   output << "4 1 5"<<endl;
+   output << "4 2 6"<<endl;
+   output << "4 3 7"<<endl;
+   output << "4 1 8"<<endl;
+   output << "4 0 9"<<endl;
+   output << endl;
+
+   // Vertices
+   output << "vertices" << endl;
+   output << 10 << endl;
+   output << endl;
+
+   // Patches
+   output<<"patches"<<endl;
+   output<<endl;
+
+   output << "# Patch 0 " << endl;
+   patch0.Print(output); output<<endl;
+   output << "# Patch 1 " << endl;
+   patch1.Print(output); output<<endl;
+   output << "# Patch 2 " << endl;
+   patch2.Print(output); output<<endl;
+   output << "# Patch 3 " << endl;
+   patch3.Print(output); output<<endl;
+   output << "# Patch 4 " << endl;
+   patch4.Print(output); output<<endl;
+
+   // Close
+   output.close();
+   delete kv0;
+   delete kv1;
+   delete kv2;
+   delete kv3;
+   delete kv4;
+   delete kvr;
+   delete kv_o1;
+   delete kv_o2;
+   delete xyf[0];
+   delete xyf[1];
+
+   cout << endl << "Boundary identifiers:" << endl;
+   cout << "   1   Bottom" << endl;
+   cout << "   2   Inflow" << endl;
+   cout << "   3   Top" << endl;
+   cout << "   4   Outflow" << endl;
+   cout << "   5   Foil section" << endl;
+   cout << "=========================================================="<< endl;
+   cout << "  "<< mdim <<"D mesh generated: " <<mesh_file.c_str()<< endl ;
+   cout << "=========================================================="<< endl;
+
+   // Print mesh info to screen
+   cout << "=========================================================="<< endl;
+   cout << " Attempting to read mesh: " <<mesh_file.c_str()<< endl ;
+   cout << "=========================================================="<< endl;
+   Mesh *mesh = new Mesh(mesh_file.c_str(), 1, 1);
+   mesh->PrintInfo();
+
+   // Print mesh to file for visualization
+   if (visit)
+   {
+      VisItDataCollection dc = VisItDataCollection("mesh", mesh);
+      dc.SetPrefixPath("Naca_cmesh");
+      dc.SetCycle(0);
+      dc.SetTime(0.0);
+      dc.Save();
+   }
+
+   if (visualization)
+   {
+      // Create glvis output
+      string glvis_file;
+      glvis_file.append(msh_path);
+      glvis_file.append("glvis_");
```

Comment:
It looks like the files glvis_naca-cmesh.mesh and naca-cmesh.mesh are the same, except for comments. Do both need to be output?

Reply:
I did this on purpose. In the script the mesh file is being generated and tested, otherwise the script would not really make sense. However, the option `--visualisation` for GlVis visualisation is mandatory to have. I thought it would make more sense to have an action associated with with this that actually gives a glvis visualisation file, instead of leaving it a dummy option which would not do anything.

Reply:
I see, it is necessary to write the file and read it back for constructing the mesh. This is generally necessary for NURBS meshes, because there is no constructor in the code. Writing a file is the only to construct a NURBS mesh. In the future, we should add NURBS mesh constructors to MFEM.

Reply:
I agree!

Reply:
Technically, you don't **need** to write to a file -- you can write to a `std::stringstream` and then read from it -- all of this will be in-memory. 😁

Reply:
Good point @v-dobrev, and I have done this for small meshes, e.g. a square or cube of a certain order. Even for such a simple mesh, such as a cube of order 3, it is tricky to write out all the DOFs in the right order. I plan to make some constructors to make the process more user-friendly, some day.

Reply:
@v-dobrev and @dylan-copeland These comments got a bit snowed under in others, so I notice them to late. @v-dobrev his suggestion is indeed possible. However, using this script without the option for glvis visualisation will then also remove the desired output of this script: the mesh itself. This might lead to unclarity in for user of this mesher as the expected output cannot be found. The script just does not do anything.